bfdriver/efi/efi_vmexits.cpp

```diff
@@ -0,0 +1,382 @@
+            return false;
+        }
+
+        if (val & ::intel_x64::msrs::ia32_efer::lme::mask) {
+            val |= ::intel_x64::msrs::ia32_efer::lma::mask;
+            ::intel_x64::vmcs::guest_cr0::set((0x60000030ULL) | (1ULL << 0) | (1ULL << 31) | (1ULL << 16));
+            ::intel_x64::vmcs::vm_entry_controls::ia_32e_mode_guest::enable();
+            ::intel_x64::vmcs::secondary_processor_based_vm_execution_controls::unrestricted_guest::disable();
+            ::intel_x64::vmcs::secondary_processor_based_vm_execution_controls::enable_ept::disable();
+            ::intel_x64::vmcs::primary_processor_based_vm_execution_controls::monitor_trap_flag::disable();
+        }
+
+        ::intel_x64::vmcs::guest_ia32_efer::set(val);
+        return advance(vmcs);
+    }
+
+    return false;
+}
+
+static bool
+handle_ept_violation(gsl::not_null<vmcs_t *> vmcs)
+{
+    ::intel_x64::vmcs::secondary_processor_based_vm_execution_controls::enable_ept::disable();
+    return true;
+}
+
+static bool
+handle_wrcr0(gsl::not_null<vmcs_t *> vmcs, control_register::info_t &info)
+{
+    // only need access type 0 but eapis doesn't handle
+    // these other access types properly when cr0 is emulated
+    auto access_type = vmcs_n::exit_qualification::control_register_access::access_type::get();
+    if (access_type == 2) {
+        auto cur = vmcs_n::guest_cr0::get();
+        info.shadow = info.shadow & ~vmcs_n::guest_cr0::task_switched::mask;
+        info.val = cur & ~vmcs_n::guest_cr0::task_switched::mask;
+    }
+    else if (access_type == 3) {
+        auto cur = vmcs_n::guest_cr0::get() & ~0xFFFFULL;
+        info.val = cur | vmcs_n::exit_qualification::control_register_access::source_data::get();
+        info.shadow &= ~0xFFFFULL;
+        info.shadow |= info.shadow | vmcs_n::exit_qualification::control_register_access::source_data::get();
+        info.val |= ::intel_x64::msrs::ia32_vmx_cr0_fixed0::get();
+    }
+    else if (access_type == 0) {
+        info.shadow = info.val;
+        info.val |= 0x30;
+    }
+    else {
+        throw std::runtime_error("handle_wrcr0 invalid access_type");
+    }
+
+    return true;
+}
+
+static bool
+handle_wrcr4(gsl::not_null<vmcs_t *> vmcs, control_register::info_t &info)
+{
+    info.shadow = info.val;
+    info.val = info.val | ::intel_x64::msrs::ia32_vmx_cr4_fixed0::get();
+
+    return true;
+}
+
+static bool
+handle_vmcall(gsl::not_null<vmcs_t *> vmcs)
+{
+    uint8_t core = thread_context_cpuid();
+    uint16_t bf = 0xFB00;
+    vmcs->save_state()->rax = (uint64_t)(bf | core);
+    return advance(vmcs);
+}
+
+static bool
+handle_init_signal(
+    gsl::not_null<vmcs_t *> vmcs)
+{
+    ::intel_x64::vmcs::guest_activity_state::set(::intel_x64::vmcs::guest_activity_state::wait_for_sipi);
+    return true;
+}
+
+static bool
+handle_sipi(gsl::not_null<vmcs_t *> vmcs)
+{
+
+    uint64_t *pml4t_p = (uint64_t *)bfvmm::memory_manager::instance()->alloc(0x1000);
```

Comment:
use the 'g_mm' macro for access to the memory manager. Also, dont use C-style casts. Use either static_cast or reinterpret_cast instead.

For other general rules-of-thumb, see the [cpp core guidelines](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines)